Video 1 - draft script (#8)
asyncapi/training: VIDEO_1_SCRIPT.md

# Video 1: What are APIs

- What are APIs
- Brief history of APIs (when and how APIs were born)
- Types of APIs
- Benefits of APIs

Hello! My name is … and today I'm going to talk about APIs. Maybe you've heard of APIs before because they're in almost all the technology we use in our everyday lives.

We're going to dive a little bit into their history and origins, explaining what they are, what types of APIs exist, and what are their benefits. Finally, we will show you some API examples to make the concept and its use clearer.

Let’s go!

## ***What are APIs***

***API is the acronym for Application Programming Interface. APIs are a tool that facilitates the communication and relationship between several applications. It’s the way apps are able to exchange messages or data and understand each other.***

APIs allow your products and services to communicate with others, without needing to know how they are implemented. An API is a set of definitions and protocols used to develop and integrate application software.

> For communicating we need to be on the same page; it's the same for machines. You see, API is an user interface that enables interaction and communication between software and a person.

Reviewer: Won't it sound better if we rephrase :

*You see, an API is a user interface that enables interaction and communication between software and a person*

Author: Yes, it sounds better this way.

As this PR is closed I made the change in the new script [file](https://github.com/Barbanio/training/blob/master/VIDEO_1_SCRIPT:_WHAT_ARE_APIS.md). 

Thanks for your feedback @thulieblack 🙂

## Benefits of APIs

Let's highlight the many benefits of APIs:

- **Adaptation**: needs change over time and APIs help anticipate changes. When working on this technology, data migration can be better supported, while offering a better review of the information. 

- **Automation**: Through APIs, workflows can be updated so that they can be done in less time while being more productive.

- **Flexibility**: enabling API access to app components provides flexibility in the delivery of services and information.  

- **Customization**: through APIs, any user can customize content and services.

- **Efficiency**: when access to an API is provided, the information is made available immediately. 

- **Integration**: APIs allow information to be easily shared between applications or products. This enables an integrated user experience.

## Example of an API

I think it would be great to give some example of API related to AsyncAPI, any idea or suggestion?

> 

Reviewer 2: Nice job! I think this will turn out a great introductory video. 
If I can give add a more general commentary about the part that talks about the history of APIs is this:
Use story-telling. Structure the information around the elements of a story: exposition, conflict, climax, and conclusion. People are interested in people. Give specific names instead of talking about companies. Who were the main contributors to the origins of APIs? What were their contributions? 
Another tip is that your story should answer what? how? why? and when?

> # Video 1: What are APIs

Reviewer 3: <img width="600" alt="Screen Shot 2022-02-03 at 6 00 48 PM" src="https://user-images.githubusercontent.com/19964402/152460159-ecd48471-3f97-4add-bfda-ce98e8654f27.png">

Question, have you ever used Grammarly Pro to help check for plagiarism? 😀👍🏽👍🏽  I currently use it to help catch myself from accidentally writing something too similar from my research sources. It can feel hard to not write too similar sometimes, or at least, to me that is how it feels. 😅  So that is why I use Grammarly Pro because it also has a feature that helps me identify how closely my own writing may match that of someone else. 

Anyways, I ran it for this Video Script 1 so far (will do so for Video Script 2 ❤️)  and I actually dismissed a few of the flagged notes because I don't think they merited a check... that said, it currently shows a 5% plagiarism rate and I thought it would help to know about the flagged sections 😀👍🏽👍🏽 🥳

Author: Wow, this's super interesting! I use Grammarly but not the Premium version. I didn't know it had this tool that I find very useful. It is difficult for me to modify and rewrite the information when it is so technical 😅

- What are APIs

> - Brief history of APIs (when and how APIs was born)

Reviewer 3: ```suggestion
- Brief history of APIs (when and how APIs were born)
```

- Types of APIs
- Benefits of APIs
- Examples of APIs

> Hello! In this video we are going to talk about APIs. I’m pretty sure you have heard more than once about them. They are everywhere! On the net, apps, and almost all the technology we use in our everyday lives. 

Reviewer 3: ```suggestion
Hello! In this video, we're going to talk about APIs. I’m pretty sure you've heard more than once about them. They're everywhere! On the net, apps, and almost all the technology we use in our everyday lives. 
```

> For doing so we are going to dive a little bit in their history and origins, explaining what they are, what types of APIs exist, what are their benefits. Finally, we will show you some examples of them to make the concept and their use clearer.

Reviewer 3: ```suggestion
We're going to dive a little bit into their history and origins, explaining what they are, what types of APIs exist, and what are their benefits. Finally, we will show you some API examples to make the concept and its use clearer.
```

Let’s go!

## What are APIs

> APIs are just the nickname of Application Programming Interfaces, a tool that facilitates the communication and relationship between two applications. It’s the way two machines are able to exchange messages or data and understand each other. 

Reviewer 3: ```suggestion
API is the acronym for Application Programming Interfaces. APIs are a tool that facilitates the communication and relationship between two applications. It’s the way two machines are able to exchange messages or data and understand each other. 
```

> APIs allow your products and services to communicate with others, without needing to know how they are implemented. After all an API could be said to be a set of definitions and protocols used to develop and integrate application software.

Reviewer 3: ```suggestion
APIs allow your products and services to communicate with others, without needing to know how they are implemented. An API is a set of definitions and protocols used to develop and integrate application software.
```

> For communicating we need to be on the same page, as machines need too.  Pretty much as a user interface enables interaction and communication between a software and a person.

Reviewer 3: ```suggestion
For communicating we need to be on the same page; it's the same for machines. You see, APIs are a user interface that enables interaction and communication between software and a person.
```

> For communicating we need to be on the same page; it's the same for machines. You see, APIs are a user interface that enables interaction and communication between software and a person.

Reviewer 4: Let's be cautious calling them "contracts" because that's what many people call the specification documents like AsyncAPI. Can be confusing later.

> Each API is designed in a particular programming language and with distinct specifications that define it (APIs may include specifications for data structures and routines, object classes or variables, from which the use of that interface is based). 
> 
> For doing things easier, it is common for each API to have complete and effective documentation (a set of tutorials, manuals and best practice rules for that programming interface).
> 
> APIs are an essential part of the development process of any application. However, it remains hidden for the end user. They are the great unknowns. The magic trick that makes things work. Undisclosed!

Reviewer 3: ```suggestion
Each API is designed in a particular programming language and with distinct specifications that define it. APIs may include specifications for data structures, routines, object classes, or variables; all based on your selected interface.  

It's easier for each API to have complete and effective documentation, such as a set of tutorials, manuals, and best practices for that programming interface.

APIs are an essential part of the development process of any application, but it remains hidden from the end-user. They are the great unknowns. The magic trick that makes things work!
```

## Brief history of APIs

> When and how APIs were born? APIs are not that old, you know? The API story took baby steps until around the beginning of the new millennium. Was then when APIs were growing fast until become the fundamental piece of the digital world where we live today. 
> 
> But, not that fast! Let’s go back to their roots: The cloud computing company Salesforce made the Internet the medium in which its service would live, based from its origins on an API.
> 
> However, the official arrival and introduction to the public of APIs took place with the dissertation written by Roy Fielding, one of the main authors of the HTTP specification on software architecture. That was a fundamental step to understanding the childhood adventures and quick growth of APIs. Their adventures, misfortunes and dreams. Their initial history, after all.
> 
> Beyond emotions, the beginning of APIs took solid steps with a certain constancy thanks to e-commerce. After all, at the end of 2000 eBay launched its application programming interface (API) together with a program that, trying to boost the area of e-commerce solutions, also encouraged APIs in the process. Moreover, eBay was not alone in this endeavor: in the summer of 2002, the giant Amazon followed suit.
> 
> Amazon put the icing on the cake with the launch of Amazon Web Services. With this platform, the company enabled developers to incorporate Amazon content and features into their own web pages, making it possible for third-party sites to search for and display the company's products.
> 
> Social networks were the next springboard where APIs gain momentum: in fact, a few months after its launch in 2004, Flickr made its platform available to developers. And that was one of the secrets of its dizzying success. Yahoo took over the social network when Flickr's REST API had quickly made it the preferred photo platform for many bloggers who could easily integrate their images into their websites.
> 
> Flickr was then ahead of its time, as it was not until two years later, in September 2006, that Facebook opened its platform for developers and its first API, which allowed access to the list of Facebook friends, photos, events and profile information. A mine of endless information. Just a month later, Twitter followed suit.
> 
> And while the social platforms were experimenting in the promising world of APIs, Google was going one step further and preparing the way to integrate its tools into third-party sites: just six months after the launch of Google Maps, the Mountain View company launched the mapping service API. They did so in response to the countless fake platforms that emerged based on hacking information from Google's tool.
> 
> After the success of its first attempt, Amazon took its APIs to the next level. Far from focusing exclusively on e-commerce, the company opted to create an entire infrastructure to support developers based on REST APIs: first with Amazon S3, to store data in the cloud, and later with Amazon EC2, which also allows developers to have virtual servers in their data centers.
> 
> Since 2010 the prevalence of the public API has skyrocketed. Mobiles finally brought APIs to our daily lives, every day, every hour, making life easier for every user, coming and going, understanding the world with the help of machines communicating each other fluently and being close to men and women, just as close as one click or quick look. 

Reviewer 3: ```suggestion
Have you ever wondered when and how APIs were born? Did you know that APIs are not that old? The API story took baby steps until around the beginning of the new millennium. That's when APIs began growing fast until they became a fundamental piece of the digital world where we live today. 

But, not that fast! Let’s go back to their roots: The cloud computing company Salesforce made the Internet the medium in which its service would live, based from its origins on an API. 

However, the official arrival and introduction to the public of APIs took place with the dissertation written by Roy Fielding, one of the main authors of the HTTP specification on software architecture. That was a fundamental step to understanding the childhood adventures and quick growth of APIs. Their adventures, misfortunes, and dreams. Their initial history, after all.

Beyond emotions, the beginning of APIs took solid steps with a certain constancy thanks to e-commerce. At the end of 2000, eBay launched its application programming interface (API) together with a program that, trying to boost the area of e-commerce solutions, also encouraged APIs in the process. Moreover, eBay was not alone in this endeavor: in the summer of 2002, the giant Amazon followed suit.

Amazon put the icing on the cake with the launch of Amazon Web Services. With this platform, the company enabled developers to incorporate Amazon content and features into their own web pages, making it possible for third-party sites to search for and display the company's products.

Social networks were the next springboard where APIs gain momentum: in fact, a few months after its launch in 2004, Flickr made its platform available to developers. And that was one of the secrets of its dizzying success. Yahoo took over the social network when Flickr's REST API quickly made it the preferred photo platform for many bloggers who could easily integrate their images into their websites.

Flickr was then ahead of its time, as it was not until two years later, in September 2006, that Facebook opened its platform for developers and its first API, which allowed access to the list of Facebook friends, photos, events, and profile information. A mine of endless information. Just a month later, Twitter followed suit.

And while the social platforms were experimenting in the promising world of APIs, Google was going one step further and preparing the way to integrate its tools into third-party sites: just six months after the launch of Google Maps, the Mountain View company launched the mapping service API. They did so in response to the countless fake platforms that emerged based on hacking information from Google's tool.

After the success of its first attempt, Amazon took its APIs to the next level. Far from focusing exclusively on e-commerce, the company opted to create an entire infrastructure to support developers based on REST APIs: first with Amazon S3, to store data in the cloud, and later with Amazon EC2, to have virtual servers in data centers.

Since 2010, the prevalence of the public API has skyrocketed. Mobiles finally brought APIs to our daily lives, every day, every hour, making life easier for every user that was coming and going. We now had support understanding the world with the help of machines that communicated with each other fluently. All this information is within grasp to men and women, merely one-click away. 
```

> E-commerce, Social media, cloud computing, mobility… Definitely, and without no doubts, we are currently living in the Web API era.

Reviewer 3: ```suggestion
E-commerce, Social media, cloud computing, mobility… Without a doubt, we are definitely living in the Web API era.
```

## Types of APIs

There are 4 types of APIs: 

> - **Web services APIs**: these are the application development interfaces that enable the exchange of information between a web service (software that provides access to a particular service via a URL) and an application. 
> 
>     There are four types of Web services APIs:
> 
>     •	SOAP (Simple Object Access Protocol), a standard protocol for exchanging information and data in XML between two objects.
> 
>     •	XML-RPC, a remote procedure call protocol that uses XML as the data format and HTTP calls as the communication system.
> 
>     • JSON-RPC, the same protocol but in JSON format. 
> 
>     •	REST (Representational State Transfer), a software architecture for hypermedia systems on the World Wide Web; a REST API uses the HTTP protocol.

Reviewer 3: ```suggestion
- **Web services APIs**: these are the application development interfaces that enable the exchange of information between an application and a web service. (A web service is a software that provides access to a particular service via a URL). 

    There are four types of Web services APIs:

    •	SOAP (Simple Object Access Protocol), a standard protocol for exchanging information and data in XML between two objects.

    •	XML-RPC, a remote procedure call protocol that uses XML as the data format and HTTP calls as the communication system.

    • JSON-RPC, the same protocol as our previously explained XML-RPC, but in JSON format. 

    •	REST (Representational State Transfer), a software architecture for hypermedia systems on the World Wide Web; a REST API that uses the HTTP protocol.
```

> - **Library-based APIs**: these types of APIs are those that allow an application to import a library from other software to exchange information. Nowadays, most of the libraries that provide access to products and services are designed in JavaScript.

Reviewer 3: ```suggestion
- **Library-based APIs**: these types of APIs are those that allow an application to import a library from other software, in order to exchange information. Nowadays, many of the libraries that provide access to products and services are designed in JavaScript.
```

> - **Library-based APIs**: these types of APIs are those that allow an application to import a library from other software, in order to exchange information. Nowadays, many of the libraries that provide access to products and services are designed in JavaScript.

Reviewer 5: `Nowadays, many of the libraries that provide access to products and services are designed in JavaScript` it is pretty controversial. I'm JS dev and fan but you know, there are plenty of haters out there ...

Maybe something like. `These are like APIs designed for a specific programming language, libraries designed for languages like JavaScript or Go` ?

> - **Class-based APIs**: this type of application development interfaces allows the connection with data around classes, as is usual in object-oriented programming with Java. The Java API uses abstract classes for the creation of applications just like any program developed in this language. These classes provide everything needed to perform all kinds of functions within these applications. The Java development interface is organized into packages and each of those packages in turn contains a set of related classes.  

Reviewer 3: ```suggestion
- **Class-based APIs**: this type of application development interface allows the connection with data around classes, as is usual in object-oriented programming with Java. The Java API uses abstract classes for the creation of applications just like any program developed in this language. These classes provide everything needed to perform all kinds of functions within these applications. The Java development interface is organized into packages and each of those packages, in turn, contains a set of related classes.  
```

> - **Class-based APIs**: this type of application development interface allows the connection with data around classes, as is usual in object-oriented programming with Java. The Java API uses abstract classes for the creation of applications just like any program developed in this language. These classes provide everything needed to perform all kinds of functions within these applications. The Java development interface is organized into packages and each of those packages, in turn, contains a set of related classes.  

Reviewer 5: is this really how people talk about this? for me it is just another type of library API

> - **Function APIs in operating systems**: software programs are continuously interacting with operating systems. In many cases, the way they do so is through APIs. Operating systems such as Windows have APIs that enable such communication between programs and the OS.

Reviewer 3: ```suggestion
- **Function APIs in operating systems**: software programs are continuously interacting with operating systems OS). In many cases, the way they do so is through APIs. Operating systems such as Windows have APIs that enable such communication between programs and the OS.
```

> ## Benefits os APIs

Reviewer 3: ```suggestion
## Benefits of APIs
```

> There are many benefits of APIs, it can be highlighted:

Reviewer 3: ```suggestion
Let's highlight the many benefits of APIs:
```

> - **Adaptation**: needs change over time and APIs help to anticipate changes. When working on this technology, data migration can be better supported, while offering a better review of the information. 

Reviewer 3: ```suggestion
- **Adaptation**: needs change over time and APIs help anticipate changes. When working on this technology, data migration can be better supported, while offering a better review of the information. 
```

- **Automation**: APIs allow IT teams rather than people to do the work. Through APIs, workflows can be updated so that they can be done in less time while being more productive.

> - **Application**: enabling API access to app components provides flexibility in the delivery of services and information.  

Reviewer 4: Shouldn't then be Flexibility instead of Application?

```suggestion
- **Flexibility**: enabling API access to app components provides flexibility in the delivery of services and information.  
```

- **Customization**: through APIs, any user can customize content and services.

> - **Efficiency**: when access to an API is provided, the content that is generated can be automatically published and made available to all channels. It allows it to be shared and distributed more easily. 

Reviewer 4: ```suggestion
- **Efficiency**: when access to an API is provided, the information is made available immediately. 
```

> - **Integration**: APIs allow content to be more easily embedded from any site or application. This ensures seamless information delivery and an integrated user experience.

Reviewer 4: ```suggestion
- **Integration**: APIs allow information to be easily shared between applications or products. This enables an integrated user experience.
```

## Example of an API

> I think it would be great to give some example of API related to AsyncAPI, any idea or suggestion?

Reviewer 3: Hmm.. the problem is that this script is already long, so a demo of an example API seems like a video of its own. 🤔  Unless this video is 10+ mins long, I guess I'm unsure as to what the goal for the 1st video is 😀

Author: The idea of the first video is to inform about what an API is and to make it a video with complete information about it. The proposal is a video lasting at most 15 minutes and then make a short version. 

I don't know how long it takes to make a demo of an API 😅. We could include it in this script and make it a video of about 15 minutes and then make two teasers. One with the theoretical part and one with the practical part. What do you think? Do you think it is better to do it in a separate script? or should the theoretical script be shorter?

Reviewer 3: got it! for sure, I am sure one of the other contributors can contribute the API demo and plan it to only be a few mins long :)

Hmm, if the API demo is only a few mins long, (say 3mins), then I would include it in both the short and long version. But anything longer than that and I would prob make two separate short videos... one could be the conceptual one and the other could be just a standalone demo... 

Curious to see what others end up thinking or suggesting too!

Author: Sounds good, it makes sense.

Reviewer 4: I agree it's already long and packed with lots of information. I think a demo may deserve its own video so we're not that constrained. Especially, because many (most?) people learn by example.

> Farewell:
> 
> I hope that the idea of what an API is and its environment is clearer now to you. It’s good to know what was the long and winding road and previous efforts done before a tool is implemented. 
> 
> If you have any questions about it do not hesitate to ask through the AsyncAPI Slack or through GitHub. Join the community and participate! Join us in the following videos :)

Reviewer 3: ```suggestion

I hope that the idea of what an API is and its environment are clearer now to you. It’s good to know what was the long, winding road and previous efforts made before a tool is implemented. 

If you have any questions about today's video, do not hesitate to ask through the AsyncAPI Slack or through our GitHub Discussions. Join our Open-Source community and participate! 
```
